File: engine/src/main/java/de/hanno/hpengine/util/Parentable.java
package de.hanno.hpengine.util;

import java.util.List;

public interface Parentable <T extends Parentable> {
    List<T> getChildren();
    default void addChildren(List<T> children) { getChildren().addAll(children); }
    default T addChildRelationship(T child) {
        child.setParent(this);
        addChild(child);
        return child;
    }
    default T addChild(T child) {
        if(!getChildren().contains(child)) {
            getChildren().add(child);
        }
        return child;
    }

    void setParent(T parent);
    default void removeParent() {
        getParent().getChildren().remove(this);
        setParent(null);
    }
    T getParent();
    default boolean hasParent() {
        return getParent() != null;
    }

    default boolean hasChildren() {
        return !getChildren().isEmpty();
    }
}

File: engine/src/main/kotlin/de/hanno/hpengine/util/Parentable.kt
package de.hanno.hpengine.util

interface Parentable<T : Parentable<T>> {
    val children: List<T>

    fun addChild(child: T)
    fun removeChild(child: T)
    fun hasChildInHierarchy(child: T): Boolean = children.contains(child) || children.any { it.hasChildInHierarchy(child) }

    var parent: T?
    val hasParent: Boolean
        get() = parent != null

    fun hasChildren(): Boolean = children.isNotEmpty()
    fun removeParent() {
        parent = null
    }
}